### PR TITLE
ADD and DROP SPARSE for T-SQL

### PR DESCRIPTION
**Pull request by Ozzah**
**Add lexer definitions for implicit tokens for T-SQL:**
ANTLR4 is spitting a bunch of warnings about implicit tokens, e.g. `PARSEONLY`, `SHOWPLAN_ALL`, etc., so I explicitly defined these in the lexer file.
I also noticed a parser node `NOEXEol` which I assume should be `NOEXEC`. I found no reference to the former in the Microsoft documentation.

**Define `IPV4_ADDR` and `IPV6_ADDR` components in T-SQL lexer:**
`IPV4_ADDR` and `IPV6_ADDR` are now defined in terms of `IPV4_OCTET` and `IPV6_QUARTET`. There may be a more elegant way to achieve this - I did try using the regex `{1-3}` syntax, however the C# code generation was invalid.

**Add `ADD SPARSE` and `DROP SPARSE` column modifiers to T-SQL parser**
**Update T-SQL examples to show `ADD SPARSE` and `DROP SPARSE`**
See: https://docs.microsoft.com/en-us/sql/t-sql/statements/alter-table-transact-sql?view=sql-server-ver15

To convert the `C4` non-sparse column to a sparse column, execute the following statement.
```sql
ALTER TABLE T1
ALTER COLUMN C4 ADD SPARSE ;
GO
```

To convert the `C4` sparse column to a nonsparse column, execute the following statement.
```sql
ALTER TABLE T1
ALTER COLUMN C4 DROP SPARSE ;
GO
```

### CODE REVIEW

**Comment by Ozzah** [comment]
What's going on with these failed CI tests? Same story with every other PR. Is there something broken in the pipeline?

**Comment by Ozzah** [comment]
Sorry, never mind. There was in fact a bug in `1928265` commit

**Comment by Ozzah** [comment]
I went ahead and added the remaining column modifiers from https://docs.microsoft.com/en-us/sql/t-sql/statements/alter-table-transact-sql?view=sql-server-ver15

```
      | { ADD | DROP } { ROWGUIDCOL | PERSISTED | NOT FOR REPLICATION | SPARSE | HIDDEN }
      | { ADD | DROP } MASKED [ WITH ( FUNCTION = ' mask_function ') ]
```

File: sql/tsql/TSqlLexer.g4 [hunk 1]
```diff
@@ -899,6 +918,12 @@ BINARY:              '0' 'X' HEX_DIGIT*;
 FLOAT:               DEC_DOT_DEC;
 REAL:                (DECIMAL | DEC_DOT_DEC) ('E' [+-]? DEC_DIGIT+);
 
+IPV4_OCTET:          (DEC_DIGIT | (DEC_DIGIT DEC_DIGIT) | ([0-2] DEC_DIGIT DEC_DIGIT));
+IPV6_QUARTET:        (HEX_DIGIT | (HEX_DIGIT HEX_DIGIT) | (HEX_DIGIT HEX_DIGIT HEX_DIGIT) | (HEX_DIGIT HEX_DIGIT HEX_DIGIT HEX_DIGIT));
```
Comment (KvanTTT) [inline sql/tsql/TSqlLexer.g4:922]: Simplify:

```suggestion
IPV4_OCTET:          DEC_DIGIT DEC_DIGIT? | [0-2] DEC_DIGIT DEC_DIGIT;
IPV6_QUARTET:        HEX_DIGIT (HEX_DIGIT (HEX_DIGIT HEX_DIGIT?)?)?;
```

File: sql/tsql/TSqlLexer.g4 [hunk 2]
```diff
@@ -107,6 +107,7 @@ GOTO:                                  'GOTO';
 GRANT:                                 'GRANT';
 GROUP:                                 'GROUP';
 HAVING:                                'HAVING';
+HIDDEN_:                               'HIDDEN';
```
Comment (KvanTTT) [inline sql/tsql/TSqlLexer.g4:110]: Probably newly added keywords should be treated as identifiers as well.

Comment (Ozzah) [reply sql/tsql/TSqlLexer.g4:110]: Sorry, this one I don't understand.

Comment (KvanTTT) [reply sql/tsql/TSqlLexer.g4:110]: I mean this word could be identifiers in some contexts, it's not a reserved keyword. It should be added to `keyword` list: https://github.com/antlr/grammars-v4/blob/master/sql/tsql/TSqlParser.g4#L4016

Comment (Ozzah) [reply sql/tsql/TSqlLexer.g4]: Ok. I've pushed an update.

**Comment by teverett** [comment]
@KvanTTT  good to pull?

File: sql/tsql/TSqlLexer.g4 [hunk 3]
```diff
@@ -899,6 +918,12 @@ BINARY:              '0' 'X' HEX_DIGIT*;
 FLOAT:               DEC_DOT_DEC;
 REAL:                (DECIMAL | DEC_DOT_DEC) ('E' [+-]? DEC_DIGIT+);
 
+IPV4_OCTET:          (DEC_DIGIT DEC_DIGIT?) | ([0-2] DEC_DIGIT DEC_DIGIT);
```
Comment (KvanTTT) [inline sql/tsql/TSqlLexer.g4:921]: Excess parentheses:

```suggestion
IPV4_OCTET:          DEC_DIGIT DEC_DIGIT? | [0-2] DEC_DIGIT DEC_DIGIT;
```

Comment (KvanTTT) [inline sql/tsql/TSqlLexer.g4:921]: Excess parentheses. Also, I guess this rule should be marked with `fragment`, see I guess it should be fragment, see https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#lexer-rules

```suggestion
fragment IPV4_OCTET:          DEC_DIGIT DEC_DIGIT? | [0-2] DEC_DIGIT DEC_DIGIT;
```

File: sql/tsql/TSqlLexer.g4 [hunk 4]
```diff
@@ -899,6 +918,12 @@ BINARY:              '0' 'X' HEX_DIGIT*;
...
+IPV6_QUARTET:        HEX_DIGIT (HEX_DIGIT (HEX_DIGIT HEX_DIGIT?)?)?;
```
Comment (KvanTTT) [inline sql/tsql/TSqlLexer.g4:922]:
```suggestion
fragment IPV6_QUARTET:        HEX_DIGIT (HEX_DIGIT (HEX_DIGIT HEX_DIGIT?)?)?;
```

File: sql/tsql/TSqlParser.g4 [hunk 5]
```diff
@@ -3047,6 +3047,16 @@ column_definition
       column_constraint*
     ;
 
+column_modifier
+    : id_ (ADD | DROP) (
+      ROWGUIDCOL
+      | PERSISTED
+      | (NOT FOR REPLICATION)
```
Comment (KvanTTT) [inline sql/tsql/TSqlParser.g4:3054]:
```suggestion
      | NOT FOR REPLICATION
```

**Review by KvanTTT — CHANGES_REQUESTED** [review]

**Review by KvanTTT — COMMENTED** [review]
It seems like all new keywords should be added to [`keyword`](https://github.com/antlr/grammars-v4/blob/master/sql/tsql/TSqlParser.g4#L4016) parse rule

```suggestion
keyword
    : ABSOLUTE
    ...
    | MASKED
    | SPARSE
    | ANSI_DEFAULTS
    | ANSI_NULL_DFLT_OFF
    | ANSI_NULL_DFLT_ON
    | ARITHIGNORE
    | FMTONLY
    | FORCEPLAN
    | IMPLICIT_TRANSACTIONS
    | NOEXEC
    | PARSEONLY
    | REMOTE_PROC_TRANSACTIONS
    | SHOWPLAN_ALL
    | SHOWPLAN_TEXT
    | SHOWPLAN_XML
    | VERSION
    | ABORT
    | HIDDEN_
    | PAUSE
```

**Comment by pguyot** [comment]
> ```antlr
> HIDDEN_
> ```

FYI, we already have 'HIDDEN' literal, it is called HIDDEN_KEYWORD.

**Comment by KvanTTT** [comment]
Could you please fix merge conflicts?

File: sql/tsql/TSqlLexer.g4 [hunk 6]
```diff
@@ -1053,6 +1066,12 @@ BINARY:              '0' 'X' HEX_DIGIT*;
 FLOAT:               DEC_DOT_DEC;
 REAL:                (DECIMAL | DEC_DOT_DEC) ('E' [+-]? DEC_DIGIT+);
 
+fragment IPV4_OCTET:          DEC_DIGIT DEC_DIGIT? | [0-2] DEC_DIGIT DEC_DIGIT;
+fragment IPV6_QUARTET:        HEX_DIGIT (HEX_DIGIT (HEX_DIGIT HEX_DIGIT?)?)?;
+
+IPV4_ADDR:           IPV4_OCTET DOT IPV4_OCTET DOT IPV4_OCTET DOT IPV4_OCTET;
+IPV6_ADDR:           IPV6_QUARTET? COLON IPV6_QUARTET? COLON IPV6_QUARTET? COLON IPV6_QUARTET? COLON IPV6_QUARTET? COLON IPV6_QUARTET? COLON IPV6_QUARTET? COLON IPV6_QUARTET?;
```
Comment (KvanTTT) [inline sql/tsql/TSqlLexer.g4:1073]: Could you please elaborate on the need of these rules? I see neither parse rules that use them nor test samples.

Comment (Ozzah) [reply sql/tsql/TSqlLexer.g4:1073]: I originally noticed on the MS SQL Server documentation there was _something_ (I can't find it anymore) that needed an IPv4/IPv6 address. But it looks like, as you said, I never made the corresponding changes to the parser rules.

I'll just remove these for now. If I eventually find the statements that have the IP addresses we can just do a separate PR for it. It doesn't seem to be related to `SPARSE`.

**Comment by KvanTTT** [comment]
Thank you for the contribution and welcome to ANTLR grammars repository!

**Pull request merged**